File: features/step_definitions/world.rb
# Step definitions for world operations

Given("w ← world\\()") do
  @w = Rayz::World.new
end

Then("w contains no objects") do
  assert_equal(@w.objects.empty?, true)
end

Then("w has no light source") do
  assert_nil(@w.light)
end

Given("s1 ← sphere\\() with:") do |table|
  @s1 = Rayz::Sphere.new
  table.raw.each do |row|
    property = row[0].strip
    value = row[1].strip

    case property
    when "material.color"
      # Parse (0.8, 1.0, 0.6) format
      match = value.match(/\(([^,]+),\s*([^,]+),\s*([^)]+)\)/)
      if match
        @s1.material.color = Rayz::Color.new(red: match[1].to_f, green: match[2].to_f, blue: match[3].to_f)
      end
    when "material.diffuse"
      @s1.material.diffuse = value.to_f
    when "material.specular"
      @s1.material.specular = value.to_f
    when "material.ambient"
      @s1.material.ambient = value.to_f
    when "transform"
      # Handle transform parsing
      if value =~ /scaling\(([^,]+),\s*([^,]+),\s*([^)]+)\)/
        @s1.transform = Rayz::Transformations.scaling(x: $1.to_f, y: $2.to_f, z: $3.to_f)
      elsif value =~ /translation\(([^,]+),\s*([^,]+),\s*([^)]+)\)/
        @s1.transform = Rayz::Transformations.translation(x: $1.to_f, y: $2.to_f, z: $3.to_f)
      end
    end
  end
end

Given("s2 ← sphere\\() with:") do |table|
  @s2 = Rayz::Sphere.new
  table.raw.each do |row|
    property = row[0].strip
    value = row[1].strip

    case property
    when "material.color"
      match = value.match(/\(([^,]+),\s*([^,]+),\s*([^)]+)\)/)
      if match
        @s2.material.color = Rayz::Color.new(red: match[1].to_f, green: match[2].to_f, blue: match[3].to_f)
      end
    when "material.diffuse"
      @s2.material.diffuse = value.to_f
    when "material.specular"
      @s2.material.specular = value.to_f
    when "material.ambient"
      @s2.material.ambient = value.to_f
    when "transform"
      if value =~ /scaling\(([^,]+),\s*([^,]+),\s*([^)]+)\)/
        @s2.transform = Rayz::Transformations.scaling(x: $1.to_f, y: $2.to_f, z: $3.to_f)
      elsif value =~ /translation\(([^,]+),\s*([^,]+),\s*([^)]+)\)/
        @s2.transform = Rayz::Transformations.translation(x: $1.to_f, y: $2.to_f, z: $3.to_f)
      end
    end
  end
end

When("w ← default_world\\()") do
  @w = Rayz::World.default_world
end

Then("w.light = light") do
  assert_equal(@w.light, @light)
end

Then("w contains s1") do
  assert(@w.objects.include?(@s1))
end

Then("w contains s2") do
  assert(@w.objects.include?(@s2))
end

When("xs ← intersect_world\\(w, r)") do
  @xs = @w.intersect(@r)
end

Given("shape ← the first object in w") do
  @shape = @w.objects[0]
end

Given("shape ← the second object in w") do
  @shape = @w.objects[1]
end

When("i ← intersection\\({float}, shape)") do |t|
  @i = Rayz::Intersection.new(t: t, object: @shape)
end

When(/^i ← intersection\((-?√\d+(?:\/\d+)?), shape\)$/) do |t_str|
  # Handle √ notation like √2 or √2/2
  t = if t_str =~ /^(-?)√(\d+)(?:\/(\d+))?$/
    sign = ($1 == "-") ? -1 : 1
    sqrt_val = Math.sqrt($2.to_f)
    denominator = ($3) ? $3.to_f : 1.0
    sign * sqrt_val / denominator
  else
    t_str.to_f
  end
  @i = Rayz::Intersection.new(t: t, object: @shape)
end

When("comps ← prepare_computations\\(i, r)") do
  @comps = @i.prepare_computations(@r)
end

When("c ← shade_hit\\(w, comps)") do
  @c = @w.shade_hit(@comps)
end

When("w.light ← point_light\\(point\\({float}, {float}, {float}), color\\({float}, {float}, {float}))") do |px, py, pz, cr, cg, cb|
  position = Rayz::Point.new(x: px, y: py, z: pz)
  intensity = Rayz::Color.new(red: cr, green: cg, blue: cb)
  @w.light = Rayz::PointLight.new(position: position, intensity: intensity)
end

When("c ← color_at\\(w, r)") do
  @c = @w.color_at(@r)
end

Given("outer ← the first object in w") do
  @outer = @w.objects[0]
end

Given("outer.material.ambient ← {float}") do |value|
  @outer.material.ambient = value
end

Given("inner ← the second object in w") do
  @inner = @w.objects[1]
end

Given("inner.material.ambient ← {float}") do |value|
  @inner.material.ambient = value
end

Then("c = inner.material.color") do
  assert_equal(@c, @inner.material.color)
end

Then("is_shadowed\\(w, p) is false") do
  assert_equal(@w.is_shadowed?(@p), false)
end

Then("is_shadowed\\(w, p) is true") do
  assert_equal(@w.is_shadowed?(@p), true)
end

Given("s1 ← sphere\\()") do
  @s1 = Rayz::Sphere.new
end

Given("s1 is added to w") do
  @w.objects << @s1
end

Given("s2 is added to w") do
  @w.objects << @s2
end

When("i ← intersection\\({float}, s2)") do |t|
  @i = Rayz::Intersection.new(t: t, object: @s2)
end

Then("c = color\\({float}, {float}, {float})") do |r, g, b|
  expected = Rayz::Color.new(red: r, green: g, blue: b)
  assert(@c == expected, "Expected #{expected.inspect} but got #{@c.inspect}")
end

When("color ← reflected_color\\(w, comps)") do
  @color = @w.reflected_color(@comps)
end

When("color ← reflected_color\\(w, comps, {int})") do |remaining|
  @color = @w.reflected_color(@comps, remaining)
end

When("color ← refracted_color\\(w, comps, {int})") do |remaining|
  @c = @w.refracted_color(@comps, remaining)
end

When("c ← refracted_color\\(w, comps, {int})") do |remaining|
  @c = @w.refracted_color(@comps, remaining)
end

When("color ← shade_hit\\(w, comps)") do
  @color = @w.shade_hit(@comps)
end

When("color ← shade_hit\\(w, comps, {int})") do |remaining|
  @color = @w.shade_hit(@comps, remaining)
end

Then("color = color\\({float}, {float}, {float})") do |r, g, b|
  expected = Rayz::Color.new(red: r, green: g, blue: b)
  assert(@color == expected, "Expected #{expected.inspect} but got #{@color.inspect}")
end

Given(/([a-z]+) ← (sphere|plane)\(\) with:/) do |var_name, shape_type, table|
  shape = case shape_type
  when "sphere"
    Rayz::Sphere.new
  when "plane"
    Rayz::Plane.new
  end

  table.raw.each do |row|
    property = row[0].strip
    value = row[1].strip

    case property
    when "transform"
      if value =~ /translation\(([^,]+),\s*([^,]+),\s*([^)]+)\)/
        shape.transform = Rayz::Transformations.translation(x: $1.to_f, y: $2.to_f, z: $3.to_f)
      elsif value =~ /scaling\(([^,]+),\s*([^,]+),\s*([^)]+)\)/
        shape.transform = Rayz::Transformations.scaling(x: $1.to_f, y: $2.to_f, z: $3.to_f)
      end
    when "material.reflective"
      shape.material.reflective = value.to_f
    when "material.transparency"
      shape.material.transparency = value.to_f
    when "material.refractive_index"
      shape.material.refractive_index = value.to_f
    when "material.color"
      match = value.match(/\(([^,]+),\s*([^,]+),\s*([^)]+)\)/)
      if match
        shape.material.color = Rayz::Color.new(red: match[1].to_f, green: match[2].to_f, blue: match[3].to_f)
      end
    when "material.ambient"
      shape.material.ambient = value.to_f
    when "material.diffuse"
      shape.material.diffuse = value.to_f
    when "material.specular"
      shape.material.specular = value.to_f
    end
  end

  instance_variable_set("@#{var_name}", shape)
end

Given(/([a-z]+) is added to w/) do |var_name|
  obj = instance_variable_get("@#{var_name}")
  @w.objects << obj
end

Given("shape.material.ambient ← {int}") do |value|
  @shape.material.ambient = value
end

Given(/^shape\.material\.transparency ← ([0-9.]+)$/) do |value|
  @shape.material.transparency = value.to_f
end

Given(/^shape\.material\.refractive_index ← ([0-9.]+)$/) do |value|
  @shape.material.refractive_index = value.to_f
end

Given(/([A-Z]) ← the first object in w/) do |var_name|
  instance_variable_set("@#{var_name}", @w.objects[0])
end

Given(/([A-Z]) ← the second object in w/) do |var_name|
  instance_variable_set("@#{var_name}", @w.objects[1])
end

Given(/([A-Z])\.material\.ambient ← ([0-9.]+)/) do |var_name, value|
  obj = instance_variable_get("@#{var_name}")
  obj.material.ambient = value.to_f
end

Given(/([A-Z])\.material\.pattern ← test_pattern\(\)/) do |var_name|
  obj = instance_variable_get("@#{var_name}")
  obj.material.pattern = Rayz::TestPattern.new
end

Given(/([A-Z])\.material\.transparency ← ([0-9.]+)/) do |var_name, value|
  obj = instance_variable_get("@#{var_name}")
  obj.material.transparency = value.to_f
end

Given(/([A-Z])\.material\.refractive_index ← ([0-9.]+)/) do |var_name, value|
  obj = instance_variable_get("@#{var_name}")
  obj.material.refractive_index = value.to_f
end

Then("color_at\\(w, r) should terminate successfully") do
  # Just call it and make sure it doesn't hang or error
  @c = @w.color_at(@r)
  assert(@c.is_a?(Rayz::Color))
end

File: python/features/steps/world.py
from behave import given, then, use_step_matcher, when

from rayz.color import Color
from rayz.intersection import Intersection, prepare_computations
from rayz.math_parser import parse_math
from rayz.point_light import PointLight
from rayz.sphere import Sphere
from rayz.transformations import scaling, translation
from rayz.tuple import Point
from rayz.world import World, default_world

use_step_matcher("re")

_V = r"([A-Za-z][A-Za-z0-9_]*)"
_A = r"([^\s,)]+)"

_TRANSFORM_MAP = {"scaling": scaling, "translation": translation}


def _apply_table(obj, table):
    import re

    # behave treats the first table row as column headers; include it as data
    all_rows = [(table.headings[0], table.headings[1])] + [(row[0], row[1]) for row in table]
    for prop, raw in all_rows:
        prop, raw = prop.strip(), raw.strip()
        if prop == "transform":
            m = re.fullmatch(r"(scaling|translation)\((.+)\)", raw)
            func = _TRANSFORM_MAP[m.group(1)]
            args = [parse_math(a.strip()) for a in m.group(2).split(",")]
            obj.set_transform(func(*args))
        elif prop == "material.color":
            m = re.fullmatch(r"\((.+),\s*(.+),\s*(.+)\)", raw)
            obj.material.color = Color(parse_math(m.group(1)), parse_math(m.group(2)), parse_math(m.group(3)))
        elif prop == "material.diffuse":
            obj.material.diffuse = parse_math(raw)
        elif prop == "material.specular":
            obj.material.specular = parse_math(raw)
        elif prop == "material.ambient":
            obj.material.ambient = parse_math(raw)
        elif prop == "material.reflective":
            obj.material.reflective = parse_math(raw)
        elif prop == "material.transparency":
            obj.material.transparency = parse_math(raw)
        elif prop == "material.refractive_index":
            obj.material.refractive_index = parse_math(raw)


# ---------------------------------------------------------------------------
# Given
# ---------------------------------------------------------------------------


@given(rf"{_V} ← world\(\)")
def step_given_world(context, var):
    setattr(context, var, World())


@given(rf"{_V} ← sphere\(\) with:")
def step_given_sphere_with_table(context, var):
    s = Sphere()
    _apply_table(s, context.table)
    setattr(context, var, s)


@given(rf"{_V} is added to {_V}")
def step_given_shape_added_to_world(context, shape_var, world_var):
    getattr(context, world_var).objects.append(getattr(context, shape_var))


@given(rf"{_V}\.light ← point_light\(point\({_A},\s*{_A},\s*{_A}\),\s*color\({_A},\s*{_A},\s*{_A}\)\)")
def step_given_world_light(context, world_var, px, py, pz, ir, ig, ib):
    getattr(context, world_var).light = PointLight(
        Point(parse_math(px), parse_math(py), parse_math(pz)),
        Color(parse_math(ir), parse_math(ig), parse_math(ib)),
    )


@given(rf"{_V} ← the first object in {_V}")
def step_given_first_object(context, var, world_var):
    setattr(context, var, getattr(context, world_var).objects[0])


@given(rf"{_V} ← the second object in {_V}")
def step_given_second_object(context, var, world_var):
    setattr(context, var, getattr(context, world_var).objects[1])


@given(rf"{_V}\.material\.ambient ← {_A}")
def step_given_shape_ambient(context, var, val):
    getattr(context, var).material.ambient = parse_math(val)


@given(rf"{_V} ← intersection\({_A},\s*{_V}\)")
def step_given_intersection(context, var, t_val, obj_var):
    setattr(context, var, Intersection(parse_math(t_val), getattr(context, obj_var)))


# ---------------------------------------------------------------------------
# When
# ---------------------------------------------------------------------------


@given(rf"{_V} ← default_world\(\)")
def step_given_default_world(context, var):
    setattr(context, var, default_world())


@when(rf"{_V} ← default_world\(\)")
def step_when_default_world(context, var):
    setattr(context, var, default_world())


@when(rf"xs ← intersect_world\({_V},\s*{_V}\)")
def step_when_intersect_world(context, world_var, ray_var):
    context.xs = getattr(context, world_var).intersect_world(getattr(context, ray_var))


@when(rf"comps ← prepare_computations\({_V},\s*{_V}\)")
def step_when_prepare_computations(context, i_var, ray_var):
    context.comps = prepare_computations(getattr(context, i_var), getattr(context, ray_var))


@when(rf"c ← shade_hit\({_V},\s*comps\)")
def step_when_shade_hit(context, world_var):
    context.c = getattr(context, world_var).shade_hit(context.comps)


@when(rf"c ← color_at\({_V},\s*{_V}\)")
def step_when_color_at(context, world_var, ray_var):
    context.c = getattr(context, world_var).color_at(getattr(context, ray_var))


@when(rf"color ← reflected_color\({_V},\s*comps\)")
def step_when_reflected_color(context, world_var):
    context.color = getattr(context, world_var).reflected_color(context.comps)


# ---------------------------------------------------------------------------
# Then
# ---------------------------------------------------------------------------


@then(r"w contains no objects")
def step_then_no_objects(context):
    assert context.w.objects == []


@then(r"w has no light source")
def step_then_no_light(context):
    assert context.w.light is None


@then(rf"{_V}\.light = {_V}")
def step_then_world_light(context, world_var, light_var):
    assert getattr(context, world_var).light == getattr(context, light_var)


@then(rf"{_V} contains {_V}")
def step_then_world_contains(context, world_var, shape_var):
    world = getattr(context, world_var)
    shape = getattr(context, shape_var)
    assert any(obj.transform == shape.transform and obj.material == shape.material for obj in world.objects)


@then(rf"is_shadowed\({_V},\s*{_V}\) is (true|false)")
def step_then_is_shadowed(context, world_var, point_var, expected):
    result = getattr(context, world_var).is_shadowed(getattr(context, point_var))
    assert result == (expected == "true")


@then(rf"c = {_V}\.material\.color")
def step_then_c_eq_material_color(context, shape_var):
    assert context.c == getattr(context, shape_var).material.color


@then(rf"color = color\({_A},\s*{_A},\s*{_A}\)")
def step_then_color_eq(context, r, g, b):
    expected = Color(parse_math(r), parse_math(g), parse_math(b))
    assert context.color == expected, f"{context.color!r} != {expected!r}"
